core/src/main/java/nl/nn/adapterframework/configuration/classloaders/ClassLoaderBase.java

```diff
@@ -237,34 +245,66 @@ protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundE
 			throw new IllegalArgumentException("classname to load may not be null");
 		}
 
+		//This is required because when using an external WebAppClassloader (ClassPath) inner classes may be retrieved from the wrong ClassLoader
+		int dollar = name.lastIndexOf("$");
+		if(dollar > 0) {
+			String baseClass = name.substring(0, dollar);
+			if(loadedCustomClasses.contains(baseClass)) {
+				return defineClass(name, resolve);
+			}
+		}
+
 		Throwable throwable = null;
 		try {
-			return getParent().loadClass(name); // First try to load the class natively
-		} catch (Throwable t) { // Catch NoClassDefFoundError and ClassNotFoundExceptions
+			return super.loadClass(name, resolve); // First try to load the class natively
+		} catch (ClassNotFoundException | NoClassDefFoundError t) { // Catch NoClassDefFoundError and ClassNotFoundExceptions
 			throwable = t;
 		}
 
-		String path = name.replace(".", "/")+".class";
-		URL url = null;
-		if(allowCustomClasses) {
-			if(log.isTraceEnabled()) log.trace(String.format("attempting to load custom class [%s] path [%s]", name, path));
-
-			url = getResource(path);
-			if(url != null && log.isDebugEnabled()) log.debug(String.format("loading custom class url [%s] from classloader [%s]", url, this.toString()));
-		} else {
-			url = getParent().getResource(path); //only allow custom code to be on the actual jvm classpath and not in a config
+		try {
+			return defineClass(name, resolve);
+		} catch (ClassNotFoundException e) {
+			e.addSuppressed(throwable);
+			throw e;
 		}
+	}
 
-		if(url != null) {
-			try {
-				byte[] bytes = Misc.streamToBytes(url.openStream());
-				return defineClass(name, bytes, 0, bytes.length);
-			} catch (Exception e) {
-				throw new ClassNotFoundException("failed to load class ["+path+"] in classloader ["+this.toString()+"]", e);
+	/**
+	 * This method will only be called for classes that have not been previously loaded yet.
+	 * Custom code will not update if you change the configuration.
+	 * 
+	 * Introspector#findExplicitBeanInfo/BeanInfoFinder#find attempts to lookup classes with the 'BeanInfo' postfix.
+	 * Introspector#findCustomizerClass attempts to lookup classes with the 'Customizer' postfix.
+	 */
+	private Class<?> defineClass(String name, boolean resolve) throws ClassNotFoundException {
+		if(getAllowCustomClasses()) {
```

Comment:
Perhaps use an early-exit strategy:
```
if (!getAllowCustomClasses()) {
    throw new ClassNotFoundException(...);
}
```

Reply:
Ja, ik vond het minder mooi omdat je dan 2x de Exception moet declareren (bij niet vinden van de URL) en bij het niet aan staan van `getAllowCustomClasses`. Voordeel zou wel zijn dat je duidelijkere excepties zou kunnen geven.

Reply:
Maar er staat sowieso al een andere `throw new ClassNotFoundException(...)` in.